spec/services/doc_auth_router_spec.rb

```diff
@@ -33,6 +33,107 @@
+
+        expect(results.tally['test2']).to be_within(iterations*percent_variance).of(target_value)
+      end
+
+      it 'doc_auth_vendor randomizes at 100 when set above 100' do
+        doc_auth_vendor_randomize_percent = 105
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor_randomize_percent).
+          and_return(doc_auth_vendor_randomize_percent)
+
+        results = []
+        (1..iterations).each do |_i|
+          results.push(DocAuthRouter.doc_auth_vendor(discriminator: SecureRandom.uuid))
+        end
+
+        expect(results.tally['test2']).to be(iterations)
+      end
+
+      it 'doc_auth_vendor randomizes at 0 when set below 0' do
+        doc_auth_vendor_randomize_percent = -5
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor_randomize_percent).
+          and_return(doc_auth_vendor_randomize_percent)
+
+        results = []
+        (1..iterations).each do |_i|
+          results.push(DocAuthRouter.doc_auth_vendor(discriminator: SecureRandom.uuid))
+        end
+
+        expect(results.tally['test2']).to be(nil)
+      end
+
+      it 'doc_auth_vendor is deterministic when presented with the same session id' do
+        doc_auth_vendor_randomize_percent = 50
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor_randomize_percent).
+          and_return(doc_auth_vendor_randomize_percent)
+
+        single_id = SecureRandom.uuid
+
+        results = []
+        (1..iterations).each do |_i|
+          results.push(DocAuthRouter.doc_auth_vendor(discriminator: single_id))
+        end
+
+        expect(results.tally['test2']).to be(iterations).or be(nil)
+        expect(results.tally['test1']).to be(iterations).or be(nil)
+      end
+
+      it 'doc_auth_vendor returns an exception when called without a session_id when randomized' do
+        expect { DocAuthRouter.doc_auth_vendor(discriminator: nil) }.to raise_error
+      end
+
+      it 'client returns randomized vendors when configured' do
+        doc_auth_vendor = 'acuant'
+        doc_auth_vendor_randomize_alternate_vendor = 'lexisnexis'
+        doc_auth_vendor_randomize_percent = 60
+
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor).and_return(doc_auth_vendor)
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor_randomize_alternate_vendor).
+          and_return(doc_auth_vendor_randomize_alternate_vendor)
+        allow(IdentityConfig.store).to receive(:doc_auth_vendor_randomize_percent).
+          and_return(doc_auth_vendor_randomize_percent)
+
+        results = []
+        (1..iterations).each do |_i|
+          client = DocAuthRouter.client(vendor_discriminator: SecureRandom.uuid).client
+          results.push(client.class.to_s)
+        end
+
+        target_value = iterations*(doc_auth_vendor_randomize_percent.to_f/100)
+
+        expect(results.tally['DocAuth::LexisNexis::LexisNexisClient']).
+          to be_within(iterations*percent_variance).of(target_value)
```

Comment:
I sorta worry we're setting ourselves up for some flaky test failures here, though admittedly I don't have much in terms of suggestions for an alternative to measure successful randomness.

Only thing that comes to mind is that 100 percent "randomize" would result in alternative, and 0 percent would not result in alternative.

Reply:
Yeah. I'm not wild about it. I have tests that do what you say at 0 and 100, and I have another one that will randomly come out as 100% one or the other.

But I also felt like it would be weird to not have a test of the randomness in the way it'll actually be used, which will almost never be 0 or 100 when it's turned on. I modified the iterations (now 1000) and variance (now 5%) to combat this then ran the set of tests around 100 times locally and never got a failure. That's not to say it's impossible, but I hope across 1000 iteration it's never off by more than +-50.

My feeling is: Let's put it out. If it fails errantly because of randomness I'll rewrite the two tests that are susceptible to flaking so that they are only testing 0 or 100 percent.